src/Interpreters/ArrayJoinAction.cpp

```diff
@@ -16,6 +18,34 @@ namespace ErrorCodes
     extern const int TYPE_MISMATCH;
 }
 
+const DataTypeArray * getArrayJoinDataType(const DataTypePtr & type, bool allow_map)
```

Comment:
Consider returning DataTypePtr from this function to write
```cpp
col.type = getArrayJoinDataType(src_col.type);
col.column = getArrayJoinColumn(src_col.column);
```

instead of

```cpp
col.type = getArrayJoinDataType(src_col.type)->shared_from_this();
col.column = getArrayJoinColumn(src_col.column)->getPtr();
```

Or even wrap it into the function `ColumnWithTypeAndName array_col = arrayJoinResutColumn(src_col);`

And in places where you check type you may use another function or check condition directly `typeid_cast<const DataTypeMap *>(type.get())) || typeid_cast<const DataTypeArray *>(type.get()))`

Reply:
I think it is a little trouble if `getArrayJoinDataType` return `DataTypePtr`, because you must cast `DataTypePtr` to `DataTypeArray *` in some places like `FunctionArrayJoin::getReturnTypeImpl` . 

Maybe we can add `ColumnWithTypeAndName convertArrayJoinColumn(const ColumnWithTypeAndName &)` function to wrap `shared_from_this` and `getPtr`, as you mentioned previously.